Evidence/re use annotated examples

anathomical opened the pull request:
## WHAT
Change the way that we handle "Annotated Examples" or "Hints" for Rules in Evidence.  This PR creates a new interface to create, edit, and delete Hints from a central place, rather than having to hunt down what Rule they're attached to.  In addition, instead of a set of form fields on Rules that allow you to define Hint data, it's been replaced with a drop-down selector of all hints, and a set of display data to show you exactly what that Hint contains when you select it.
## WHY
We want to be able to use the same Hint on multiple Rules (rather than the current one-to-one relationship set we maintain).  We also want to be able to make broad tweaks to those hints from a central interface.
## HOW
- Create a new Evidence CRUD resource for `Hint` objects so that they can be managed independently of `Rules`
- Modify the `Rule` model so that it can no longer accept nested `Hint` data, preventing us from accidentally impacting a `Hint` with broader implications than is obvious
- Add a migration and update `ActiveRecord` relationships to make it so that one `Hint` can be related to many `Rules` rather than the old one-to-one setup
- Create a new UI/UX for managing Hints in the Evidence dashboard
- Update Rule detail UI/UX that used to let staff create Hints in-place on those pages with a drop-down that lets staff select existing Hint objects to assign them to a Rule

### Screenshots
![image](https://user-images.githubusercontent.com/331565/194440656-b0eeda64-9ee0-407e-acd2-af33d9b4e458.png)
![image](https://user-images.githubusercontent.com/331565/194440714-aa4d4baf-fabd-487f-ad61-0efc39eaf8d2.png)
![image](https://user-images.githubusercontent.com/331565/194440821-195d1e7f-b3c1-4b1c-aa18-b9be4d0d29e1.png)

### Notion Card Links
https://www.notion.so/quill/Save-Re-use-annotated-examples-13ff1dbd4a5340d38639dd0412440644

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Yes
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

anathomical committed:
Make back-end changes to allow Hints to be used on multiple Rules

anathomical committed:
Add rake tasks to create new RuleHint records based Hints

anathomical committed:
WIP front-end for Hint management

anathomical committed:
Clean up backend code and migrations for new Hint structure

anathomical committed:
Update rake task for updated structure

anathomical committed:
Basic hint functionality changes:

- Create a new UI/UX for editing and managing hints at a global level
- Update rule edit pages to let uses pick Hints rather than create them on rule pages, so that they can be re-used

anathomical committed:
Tweak hint picker to be clearer when no hint is selected

anathomical committed:
Minor style structure tweak for dropdown label

anathomical committed:
Add path params and history to the Hints editor

anathomical committed:
Allow users to un-set Hints on a rule

anathomical committed:
Install the migration for new Hint structure from Evidence engine

anathomical committed:
Fix syntax error

anathomical committed:
Fix linting errors

anathomical committed:
Fix rubocop linting error

anathomical committed:
Remove test for removed model

anathomical committed:
Remove old code that's no longer in use

anathomical committed:
Do some clean-up on old tests for new data shape

anathomical committed:
Update front-end navigation test

brendanshean commented on services/QuillLMS/engines/evidence/app/controllers/evidence/hints_controller.rb:
```diff
@@ -0,0 +1,60 @@
+
+    # GET /hints.json
+    def index
+      @hints = Evidence::Hint
```
I think this assignment would be more consistent / conventional:

```
@hints = Evidence::Hint.all
render json: @hints
```

brendanshean commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/configureRules/ruleHintPicker.tsx:
```diff
@@ -0,0 +1,29 @@
+  onHintChange,
+}) => {
+
+  const hintId = (hint && hint.id) || null
```
What about using the safe navigation (elvis) operator for these assignments:
(e.g. `const hintId = hint?.id`)

anathomical replied:
I think you're right that safe navigation is better here, but Emilia made an even better suggestion to replace all four of these lines with an expansion call, which I think is cleaner.

brendanshean commented on services/QuillLMS/client/app/bundles/Staff/helpers/evidence/ruleHelpers.tsx:
```diff
@@ -320,7 +320,7 @@ export const buildRule = ({
     suborder: suborder ? suborder : order,
     universal: universal,
-    hint_attributes: ruleHint,
+    hint_id: (ruleHint && ruleHint.id) || null,
```
Consider using safe navigation operator.

brendanshean commented on services/QuillLMS/client/app/bundles/Staff/utils/evidence/hintAPIs.ts:
```diff
@@ -0,0 +1,70 @@
+export const fetchHints = async () => {
+  const response = await apiFetch(`hints`);
+  let hints = await response.json();
+  if(hints && hints.hints) {
```
Consider `hints?.hints`

brendanshean commented on services/QuillLMS/engines/evidence/spec/controllers/evidence/hints_controller_spec.rb:
```diff
@@ -0,0 +1,115 @@
+        get(:index)
+        parsed_response = JSON.parse(response.body)
+        expect(response.status).to eq(200)
+        expect(parsed_response.class).to(eq(Array))
```
You could collapse 15-16 into `expect(parsed_reponse).to match_array([])`

brendanshean commented on services/QuillLMS/engines/evidence/spec/controllers/evidence/rules_controller_spec.rb:
```diff
@@ -306,6 +306,14 @@ module Evidence
+      it 'should connect a Hint when a hint_id is passed' do
+        hint = create(:evidence_hint)
+
+        post(:create, :params => ({ :rule => ({ :concept_uid => rule.concept_uid, :note => rule.note, :name => rule.name, :optimal => rule.optimal, :state => rule.state, :suborder => rule.suborder, :rule_type => rule.rule_type, :universal => rule.universal, :hint_id => hint.id }) }))
```
For readability, consider breaking this line up into multiple lines.

brendanshean reviewed (approved):
Very nice work on this, The code was very readable both on frontend and backend.  I'm still learning react-query so I'm bookmarking this PR as a reference for checking understanding later.  For that reason, I would recommend at least one more approval given the amount of frontend work here.

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/EvidenceLanding.tsx:
```diff
@@ -14,6 +15,7 @@ const EvidenceLanding = () => (
       <Route component={UniversalRule} path='/universal-rules/:ruleId' />
       <Route component={UniversalRulesIndex} path='/universal-rules' />
       <Route component={Activities} path='/activities' />
+      <Route component={Hints} path='/hints/:hintId?' />
```
Should that question mark be at the end there?

anathomical replied:
So according to this StackOverflow article (https://stackoverflow.com/questions/35604617/react-router-with-optional-path-parameter), React Router doesn't document this particularly well, but the `?` after a param in a route means that the param is optional.  Doing it this way means that `#/hints` and `#/hints/123` both load the same component `Hints`.  There might be another way to do this that's more in line with our existing approaches, but I didn't see anything obvious in my survey of routes.

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/__tests__/activity.test.tsx:
```diff
@@ -12,7 +12,7 @@ describe('Activity component', () => {
     </MemoryRouter>
   );
 
-  it('should render 2 NavLinks', () => {
-    expect(container.find(NavLink).length).toEqual(3);
```
lol at this existing mismatch

anathomical replied:
I did the same thing.  It was one of those "clearly was 2 at one point" situations.

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/configureRules/ruleHint.tsx:
```diff
@@ -69,19 +63,18 @@ const RuleHint = ({
 
   return(
     // eslint-disable-next-line
     <React.Fragment>
-      {hint && !hint._destroy ? renderHint() : <button className="add-feedback-layer quill-button small primary outlined" onClick={handleHintCreation} type="button">Add Hint</button>}
+      {renderHint()}
     </React.Fragment>
```
NBD but this doesn't need to be nested in a Fragment anymore, and I don't think that `eslint-disable` is necessary any longer.

anathomical replied:
@emilia-friedberg I tried removing this wrapper, and started getting errors on render: `Uncaught Error: Objects are not valid as a React child (found: object with keys {}). If you meant to render a collection of children, use an array instead.`.  There may be a better way to get that error to stop happening besides wrapping the return value in a `React.Fragment>` element, but that did work.  Any suggestions for a better solution?

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/configureRules/ruleHintDropdown.tsx:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  return(
+    // eslint-disable-next-line
```
Why is this necessary?

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/configureRules/ruleHintPicker.tsx:
```diff
@@ -0,0 +1,29 @@
+  onHintChange,
+}) => {
+
+  const hintId = (hint && hint.id) || null
+  const hintExplanation = (hint && hint.explanation) || null
+  const hintImageLink = (hint && hint.image_link) || null
+  const hintImageAltText = (hint && hint.image_alt_text) || null
```
I think for all of these, the tidiest way to do it would just be:

`{ id, explanation, image_link, image_alt_text } = hint || {}`

You'd have to change the variable names below, but the file is so small I don't think you're at any risk of getting confused about what those attributes refer to.

anathomical replied:
That seems reasonable.

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/configureRules/ruleHintPicker.tsx:
```diff
@@ -0,0 +1,29 @@
+  const hintImageAltText = (hint && hint.image_alt_text) || null
+
+  return(
+    // eslint-disable-next-line
```
I don't think you need this here either.

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/hints.tsx:
```diff
@@ -0,0 +1,153 @@
+  const flashSnackbar = (text: string) => {
+    setSnackbarText(text)
+    setSnackbarVisible(true)
+    setTimeout(() => setSnackbarVisible(false), defaultSnackbarTimeout)
```
Just FYI, in hooks-based React components, you can use the `useSnackbarMonitor` hook to avoid some of this boilerplate.

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/hints.tsx:
```diff
@@ -0,0 +1,153 @@
+  }
+
+  const handleSaveResponse = (response) => {
+    if (response.errors.length > 0) setErrors(result.errors)
```
You can just do `if (response.errors.length)`

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/hints.tsx:
```diff
@@ -0,0 +1,153 @@
+  }
+
+  const handleSaveHint = () => {
+    if (hint.id == '') {
```
The repetition of checking against an empty string for logic makes me a little nervous - I might either set that empty string to be a constant (ex: `const NO_VALUE = ''` at the top and then `hint.id === NO_VALUE` throughout the file) or check against `hint.id.length` instead.

emilia-friedberg reviewed (approved):
This looks great. I made some clean-up suggestions on the frontend but nothing major - really nice work.

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/hints.tsx:
```diff
@@ -0,0 +1,153 @@
+  const queryClient = useQueryClient()
+
+  if (hintId && hintsData && hint.id != hintId) {
+    const hintFromUrl = hintsData.hints.find((hint) => hint.id == hintId)
```
line 37 should be `!==` and line 38 should be `===`

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/hints.tsx:
```diff
@@ -0,0 +1,153 @@
+
+  const history = useHistory();
+
+  const newHint: HintInterface = {
```
nbd, but this could be pulled out and exported from the Evidence constants file where we have similar constants for a new activity, new rule, etc: https://github.com/empirical-org/Empirical-Core/blob/develop/services/QuillLMS/client/app/constants/evidence.ts#L184

anathomical replied:
Good call-out.  I like that approach better anyway.

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/hints.tsx:
```diff
@@ -0,0 +1,153 @@
+  }
+
+  const deleteButton = () => {
+    if (hint.id == '') return;
```
lines 123 and 128 should be `===`

eadams17 reviewed (approved):
looks good! just a few, small comments

anathomical committed:
Change the dynamic key so that it doesn't update itself

When the contents of the TextArea are updated, we don't want that in itself to trigger a re-render of the component by changing its key

anathomical committed:
Make some updates based on PR feedback

anathomical committed:
Clean up linting errors

anathomical committed:
Add snackbar notification for creation and update

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into evidence/re-use-annotated-examples-for-merge
